### PR TITLE
build_requirejs

### PR DESCRIPTION
**Pull request by kaapstorm**
[Context](https://dimagi-dev.atlassian.net/browse/HI-744). TLDR: ACE / jsonwidget is broken, sometimes, when the `build_requirejs` management command is called with Python 3. (The UCR UI works, but a lot of other stuff doesn't.)

This is a draft PR to get feedback / someone to point out the totally obvious thing I'm missing / the tree I should have been barking up.

Here is the approach I took:
1. I ran `./manage.py build_requirejs --local` from a Py2 virtualenv and again from a Py3 virtualenv, and looked at a diff of the (mostly `bundle.js`) files that were generated. The files generated by Py3 were consistently bigger. The most obvious difference was that all the Py3-generated files I looked at included jquery (among other things), where only some of the Py2-generated ones did.

2. My first suspect was the warnings given about git subprocesses. They are caused by changes in the way Py3 deals with subprocesses, so I changed the script to use context managers to close subprocesses pythreenically. (Not a word? It should be.) ... This did not resolve the problem.

3. Then I refactored the script, to understand it better. (You'll see a chunk of refactors below.) Nothing else obvious jumped out at me.

4. I added breakpoints and compared values when run with Py2 vs Py3. ... They were all the same. The generated files were also the same now!

5. Confusing, but not a problem: Clearly my refactor had made Py2 behave like Py3. No worries, I can just use `git bisect` to see which refactor broke Py2-generated files.

6. `git bisect` says my very first commit broke Py2. Well that can't be! I checked after my change to use Popen as a context manager, and Py2-generated files were different from Py3-generated files.

7. As a sanity check I switch back to master. ... Now `build_requirejs` is producing identical files in both Py2 and Py3. Insanity confirmed, I went to bed.

I am going to do steps 1, 2 and 7 again, and check whether my insanity was just temporary. But if you can spot something wrong with my method or my code, I would be mighty appreciative.

### CODE REVIEW

File: corehq/apps/hqwebapp/management/commands/build_requirejs.py [hunk 1]
```diff
@@ -205,3 +97,141 @@ def handle(self, **options):
         resource_versions["hqwebapp/js/resource_versions.js"] = self.get_hash(filename)
 
         self.overwrite_resources(resource_versions)
+
+
+def _confirm_or_exit():
+    proc = subprocess.Popen(["git", "diff-files", "--ignore-submodules", "--name-only"],
+                            stdout=subprocess.PIPE)
```
Comment (millerdev) [inline corehq/apps/hqwebapp/management/commands/build_requirejs.py]: To be consistent with the first commit, should this use `with subprocess.Popen(...) as proc:`?

File: corehq/apps/hqwebapp/management/commands/build_requirejs.py [hunk 2]
```diff
@@ -205,3 +97,141 @@ def handle(self, **options):
...
+    Returns a dict of all main js modules, grouped by directory:
+    {
+        'locations/js': set(['locations/js/import', 'locations/js/location', ...  ]),
+        'linked_domain/js': set(['linked_domain/js/domain_links']),
+        ...
+    }
+    """
+    dirs = defaultdict(set)
+    for filename in html_files:
+        proc = subprocess.Popen(["grep", r"^\s*{% requirejs_main [^%]* %}\s*$", filename],
```
Comment (millerdev) [inline corehq/apps/hqwebapp/management/commands/build_requirejs.py]: To be consistent with the first commit, should this use `with subprocess.Popen(...) as proc:`?

**Review by millerdev — COMMENTED** [review]
Nothing jumped out, but didn't take a lot of time to review. I'd be happy to pair on it if you want—sounds like you have a lot of context that would be helpful.

File: corehq/apps/hqwebapp/management/commands/build_requirejs.py (hunk 1 above)
Comment (kaapstorm) [reply corehq/apps/hqwebapp/management/commands/build_requirejs.py]: I agree. I was planning to do that later -- only the git subprocesses were resulting in warnings -- so I figured that I shouldn't change any functionality in the refactor commits, and make changes afterwards.

**Comment by kaapstorm** [comment]
I am splitting this PR into two: 
- One for the fix, 
- and [one for the refactor](https://github.com/dimagi/commcare-hq/pull/24954)

**Pull request closed without merging**